src/core/sock/sock-redirect.cpp

```diff
@@ -2520,6 +2637,56 @@ extern "C" EXPORT_SYMBOL int epoll_ctl(int __epfd, int __op, int __fd, struct ep
     if (!epfd_info) {
         errno = EBADF;
     } else {
+#if defined(DEFINED_ENVOY)
+        if (g_p_app->workers_num > 0) {
+            auto iter = g_p_app->attr.envoy.map_listen_fd.find(__fd);
+            if (iter != g_p_app->attr.envoy.map_listen_fd.end()) {
+                socket_fd_api *p_socket_object = NULL;
+                p_socket_object = fd_collection_get_sockfd(__fd);
+                if (iter->second == gettid()) {
+                    /* process listen sockets from main thread */
+                    if (p_socket_object) {
+                        rc = p_socket_object->listen(p_socket_object->m_back_log);
+                        if (rc < 0) {
+                            return rc;
+                        }
+                    }
+                    std::lock_guard<decltype(g_p_app->m_lock)> lock(g_p_app->m_lock);
+                    g_p_app->attr.envoy.map_listen_fd.erase(iter);
+                } else if (__op == EPOLL_CTL_ADD) {
+                    static int worker_id = 0;
+
+                    /* original listen socket should be on*/
+                    int sleep_count = 5;
+                    while (!p_socket_object && !worker_id) {
```

Comment:
Are we waiting for worker_id to become non zero? then it should be atomic.

Reply:
This loop is needed to wait original parent fd that should be the first.
Original parent fd has not null `p_socket_object`.
All other fd should start initialization starting `worker_id>0`. It is not interesting to get actual worker_id number in this loop.